### PR TITLE
fix: recover stored value on init

### LINKED ISSUE #25: Store doesn't get saved value on init
On page refresh controller doesn't get the storage value

It should be added in the controller contructor:

```
constructor() {
   ....
   this.__value = this.__readValueFromStorage();
}
```

### PR DESCRIPTION
**Pull request by jorgecasar**
Closes #25

### CODE REVIEW

**Comment by 43081j** [comment]
had a long think about it because i wasn't sure if it makes more sense in `hostConnected`, but i think you're right putting it in the constructor actually

could you also add a test?

basically similar to this suite:
https://github.com/43081j/relit/blob/326ddeb1acb5decd8c2b8e0ef8c02e7f930c1075/src/test/controllers/localStorage_test.ts#L88-L105

it belongs in its own `describe` like that since you want to set `test-key` in your `beforeEach` (directly in local storage)

almost the same test, just don't pass a default. although it may be worth doing another where you _do_ pass a default, and assert it uses the one from storage rather than the default

File: src/test/controllers/localStorage_test.ts [hunk 1]
```diff
@@ -103,4 +103,43 @@ suite('LocalStorageController', () => {
...
+      });
+
+      test('initialises to saved value', () => {
+        const storageValue = window.localStorage.getItem('test-key');
+        assert.equal(controller.value, [7, 8]);
+        assert.equal(element.shadowRoot!.textContent, '[7,8]');
+        assert.is(storageValue, '[7,8]');
+      });
+
+      suite('with default value', () => {
```
Comment (43081j) [inline src/test/controllers/localStorage_test.ts:129]: i think you meant to close the previous suite off before this one (so both are a child of `with saved value`)

Comment (jorgecasar) [reply src/test/controllers/localStorage_test.ts:129]: No, I use the suite 'with saved value' as umbrella for with and without default value

Comment (jorgecasar) [reply src/test/controllers/localStorage_test.ts:129]: Maybe something like this is better:

```
suite('with saved value', () => {
      setup(() => {
        window.sessionStorage.setItem('test-key', JSON.stringify([7, 8]));
        element = document.createElement('test-element') as TestElement;
        element.template = () => html`${JSON.stringify(controller.value)}`;
        document.body.appendChild(element);
      });

      test('initialises to saved value', async () => {
        controller = new SessionStorageController(element, 'test-key');
        element.controllers.push(controller);
        await element.updateComplete;

        const storageValue = window.sessionStorage.getItem('test-key');
        assert.equal(controller.value, [7, 8]);
        assert.equal(element.shadowRoot!.textContent, '[7,8]');
        assert.is(storageValue, '[7,8]');
      });

      test('ignore default value and use saved value', async () => {
        controller = new SessionStorageController(element, 'test-key', [5, 6]);
        element.controllers.push(controller);
        await element.updateComplete;

        const storageValue = window.sessionStorage.getItem('test-key');
        assert.equal(controller.value, [7, 8]);
        assert.equal(element.shadowRoot!.textContent, '[7,8]');
        assert.is(storageValue, '[7,8]');
      });
    });
    ```

Comment (43081j) [reply src/test/controllers/localStorage_test.ts]: i think we're in agreement! the code you had was this structure:

- with saved value
  - sub-suite A
    - sub-suite B

i just meant it probably was meant to be:

- with saved value
  - sub-suite A
  - sub-suite B

(i.e. using `with saved value` as an umbrella like you said)

what you pushed also works though! either is fine

File: src/test/controllers/localStorage_test.ts [hunk 2]
```diff
@@ -85,22 +85,34 @@ suite('LocalStorageController', () => {
     });
   });
 
-  suite('with default value', () => {
```
Comment (43081j) [inline src/test/controllers/localStorage_test.ts:88]: sorry last one 🙈  it looks like you overwrote the old `with default value` suite here (which needs to still exist, was already in main)

basically what you have in the sessionStorage tests is right

Comment (jorgecasar) [reply src/test/controllers/localStorage_test.ts]: Sorry, copy and paste mistake 🤦

**Review by 43081j — APPROVED** [review]
looks good to me!

do love a PR which has more tests than code 😂 they will help out a lot though

**Pull request merged**